In [1]:
from neural_net import NeuralNetwork
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris
from metrics import confusion_matrix, prediction_stats, accuracy, precision, recall, f1

In [2]:
dataset = load_iris()

X = dataset.data
Y = dataset.target

In [3]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1)

In [4]:
NN = NeuralNetwork(max_iter=200, batch_size=10, error_threshold=0.1, learning_rate=0.1, random_state=42069)
NN.add_layer(n_neuron=4)
NN.add_layer(n_neuron=3, activation_function='relu')
NN.add_layer(n_neuron=2, activation_function='relu')
NN.add_layer(n_neuron=2, activation_function='sigmoid')
NN.add_layer(n_neuron=3)

In [5]:
NN.fit(X_train, Y_train)
y_pred = NN.predict(X_test)

D:\Kuliah\ML\ML-IF3270\src\C\neural_net.py:92: RuntimeWarning: invalid value encountered in multiply
  currLayer.error_term = np.sum(d_layer * wkh_dk, axis=0)


### Hasil Implementasi

In [6]:
print(f'Confusion Matrix:\n {confusion_matrix(Y_test, y_pred)}')
print(f'Precision = {precision(Y_test, y_pred)}')
print(f'Accuracy score = {accuracy(Y_test, y_pred)}')
print(f'Recall Score = {recall(Y_test, y_pred)}')
print(f'F1 Score = {f1(Y_test,y_pred)}')

Confusion Matrix:
 [[5 0 0]
 [6 0 0]
 [4 0 0]]
Precision = 0.1111111111111111
Accuracy score = 0.3422222222222222
Recall Score = 0.3333333333333333
F1 Score = 0.16666666666666666


### Hasil Sklearn

In [7]:
from sklearn.metrics import confusion_matrix, precision_score, accuracy_score, recall_score, f1_score

In [8]:
print(f'Confusion Matrix:\n {confusion_matrix(Y_test, y_pred)}')
print(f'Precision = {precision_score(Y_test, y_pred, average="weighted")}')
print(f'Accuracy score = {accuracy_score(Y_test, y_pred)}')
print(f'Recall Score = {recall_score(Y_test, y_pred, average="weighted")}')
print(f'F1 Score = {f1_score(Y_test,y_pred, average="weighted")}')

Confusion Matrix:
 [[5 0 0]
 [6 0 0]
 [4 0 0]]
Precision = 0.1111111111111111
Accuracy score = 0.3333333333333333
Recall Score = 0.3333333333333333
F1 Score = 0.16666666666666666


C:\Users\fnaab\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Tabel Perbandingan Precission, Accuracy, Recall, dan F1 Score

#### Algoritma yang Dibuat

| Jenis Score | Nilai |
| --- | --- |
| Precision | 0.1111111111111111 |
| Accuracy | 0.3422222222222222 |
| Recall | 0.3333333333333333 |
| F1 | 0.16666666666666666 |

### Algoritma Sklearn

| Jenis Score | Nilai |
| --- | --- |
| Precision | 0.1111111111111111 |
| Accuracy | 0.3333333333333333 |
| Recall | 0.3333333333333333 |
| F1 | 0.16666666666666666 |

### Analisis

In [ ]:
Hasil yang didapat antara implementasi dan dari sklearn sama persis. Dengan demikian dapat disimpulkan implement

Warning atau peringatan pada hasil dari sklearn, disebabkan terdapat kelas yang tidak muncul pada data test tetapi muncul pada hasil prediksi, ataupun sebaliknya.